src/v/archival/segment_reupload.cc

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "segment_reupload.h"
+
+#include "cloud_storage/partition_manifest.h"
+#include "logger.h"
+#include "storage/disk_log_impl.h"
+#include "storage/fs_utils.h"
+
+namespace archival {
+segment_collector::segment_collector(
+  model::offset begin_inclusive,
+  const cloud_storage::partition_manifest* manifest,
+  const storage::disk_log_impl* log,
+  size_t max_uploaded_segment_size)
+  : _begin_inclusive(begin_inclusive)
+  , _manifest(manifest)
+  , _log(log)
+  , _max_uploaded_segment_size(max_uploaded_segment_size) {}
+
+void segment_collector::collect_segments() {
+    if (_manifest->size() == 0) {
+        vlog(
+          archival_log.warn,
+          "No segments to collect for ntp {}",
+          _manifest->get_ntp());
+        return;
+    }
+
+    align_begin_offset_to_manifest();
+
+    if (_begin_inclusive >= _manifest->get_last_offset()) {
+        vlog(
+          archival_log.warn,
+          "Start offset {} is ahead of manifest last offset {} for ntp {}",
+          _begin_inclusive,
+          _manifest->get_last_offset(),
+          _manifest->get_ntp());
+        return;
+    }
+
+    do_collect();
+}
+
+segment_collector::segment_seq&& segment_collector::take_segments() {
+    return std::move(_segments);
+}
+
+void segment_collector::do_collect() {
+    auto it = _manifest->segment_containing(_begin_inclusive);
+    if (it == _manifest->end()) {
+        return;
+    }
+
+    // Crossing this boundary means that the collection can replace at least one
+    // segment in manifest.
+    auto replace_boundary = it->second.committed_offset;
+
+    auto start = _begin_inclusive;
+    model::offset current_segment_end{0};
+    size_t collected_size = 0;
+    while (current_segment_end < _manifest->get_last_offset()) {
+        auto result = find_next_compacted_segment(start);
```

Comment:
nit: would find_next_adjacent_compacted_segment be a better name? It makes it explicit that we can't skip segments.

Reply:
Compacted segments are always supposed to be adjacent on the local disk log (from my understanding of the compaction process), so they should always be adjacent. I will confirm this.